File: quant1x/data/market.go
// Copyright (c) Quant1X <wangfengxy@sina.cn>.
// Licensed under the MIT License.
//
// market — 市场/证券代码识别与纠正, 与 Python data/market.py 对齐

package data

import (
	"fmt"
	"strings"

	"github.com/quant1x/quant1x/quant1x/base"
	"github.com/quant1x/quant1x/quant1x/data/meta"
	"github.com/quant1x/quant1x/quant1x/data/meta/ticker_rules"
	"github.com/quant1x/quant1x/quant1x/data/schema"
)

// ============================================================
// Type aliases from meta package (convenience re-exports).
// ============================================================

type Exchange = meta.Exchange
type InstrumentInfo = meta.Instrument
type Timestamp = meta.Timestamp
type InstrumentType = meta.InstrumentType
type Transaction = schema.Transaction

// ============================================================
// Exchange constants from meta.
// ============================================================

const (
	SSE  = meta.SSE
	SZSE = meta.SZSE
	BSE  = meta.BSE
)

// ExchangeSSE is a deprecated alias kept for test compatibility.
const ExchangeSSE = meta.SSE

// ExchangeSZSE is a deprecated alias kept for test compatibility.
const ExchangeSZSE = meta.SZSE

// ============================================================
// Instrument type constants.
// ============================================================

const SecurityTypeUnknown = meta.InstrumentTypeUnknown

// ============================================================
// Timestamp functions (re-exported from meta).
// ============================================================

var (
	NowTimestamp           = meta.NowTimestamp
	PreMarketTimestamp     = meta.PreMarketTimestamp
	ParseTimestamp         = meta.ParseTimestamp
	LastTradingDay         = meta.LastTradingDay
	NewTimestampFromString = meta.NewTimestampFromString
	DateRange              = meta.DateRange
	ZeroTimestamp          = meta.ZeroTimestamp
	NewTimestampFromTime   = meta.NewTimestampFromTime
	CanInitialize          = meta.CanInitialize
)

// ============================================================
// String/Bytes helpers (delegated to std).
// ============================================================

// String2Bytes converts a string to a byte slice.
func String2Bytes(s string) []byte {
	return base.String2Bytes(s)
}

// Bytes2String converts a byte slice to a string.
func Bytes2String(b []byte) string {
	return base.Bytes2String(b)
}

// ============================================================
// BuildInstrument — constructs a full instrument code string.
// ============================================================

// BuildInstrument constructs a full instrument code string from exchange and ticker.
// For Chinese markets: "SH600000". For others: "AAPL.US".
func BuildInstrument(ex meta.Exchange, ticker string) string {
	if ex.Region() == meta.RegionCN {
		return fmt.Sprintf("%s%s", ex.Identifier(), ticker)
	}
	return fmt.Sprintf("%s.%s", ticker, ex.Identifier())
}

// ============================================================
// GetFirstMarketDate — earliest trading day for a given exchange.
// ============================================================

// GetFirstMarketDate returns the first market date timestamp for a given exchange.
func GetFirstMarketDate(ex meta.Exchange) Timestamp {
	switch ex {
	case meta.SSE, meta.SZSE:
		return meta.PreMarketTimestamp(1990, 12, 19)
	case meta.BSE:
		return meta.PreMarketTimestamp(2021, 11, 15)
	default:
		return meta.PreMarketTimestamp(1990, 12, 19)
	}
}

// ============================================================
// AssertIndexByMarketAndCode — checks if a code represents an index/block.
// ============================================================

// AssertIndexByMarketAndCode checks if a security code represents an index or block
// for the given exchange. Mirrors the Rust assert_index_by_market_and_code and
// the C++ assert_index_by_security_code.
//
// SSE codes starting with 000/880/881 are indices/blocks.
// SZSE codes starting with 399 are indices.
// BSE codes starting with 899 are indices.
func AssertIndexByMarketAndCode(ex meta.Exchange, code string) bool {
	if len(code) < 3 {
		return false
	}
	switch ex {
	case meta.SSE:
		return code[:3] == "000" || code[:3] == "880" || code[:3] == "881"
	case meta.SZSE:
		return code[:3] == "399"
	case meta.BSE:
		return code[:3] == "899"
	default:
		return false
	}
}

// ============================================================
// DetectInstrumentTypeByRule — 根据交易所与代码检测证券类型
// ============================================================

// DetectInstrumentTypeByRule 使用对应交易所的规则检测证券类型, 与 Python data/market.py 对齐.
// 若交易所无匹配规则, 返回 InstrumentTypeUnknown.
func DetectInstrumentTypeByRule(exchange meta.Exchange, code string) meta.InstrumentType {
	var rules []ticker_rules.CodeRule
	switch exchange {
	case meta.SSE:
		rules = ticker_rules.SseRules()
	case meta.SZSE:
		rules = ticker_rules.SzseRules()
	case meta.BSE:
		rules = ticker_rules.BseRules()
	case meta.HKEX:
		rules = ticker_rules.HkexRules()
	case meta.USA:
		rules = ticker_rules.UsaRules()
	default:
		return meta.InstrumentTypeUnknown
	}
	return ticker_rules.MatchRule(code, rules).InstrumentType
}

// ============================================================
// DetectSymbol — 检测并解析证券代码的市场及类型
// ============================================================

// DetectSymbol 检测并解析证券代码的市场类型及证券类型, 与 Python data/market.py::detect_symbol 对齐.
func DetectSymbol(inputStr string) meta.Instrument {
	s := strings.ToLower(strings.TrimSpace(inputStr))
	if s == "" {
		return meta.Instrument{Exchange: meta.UNKNOWN, Type: meta.InstrumentTypeUnknown}
	}
	pureCode := s

	var ticker string
	var exchange = meta.UNKNOWN
	var typ = meta.InstrumentTypeUnknown

	// 1. 判断前缀: sh600000
	if len(pureCode) >= 2 {
		if ex, err := meta.ParseExchange(pureCode[:2]); err == nil && ex != meta.UNKNOWN {
			ticker = pureCode[2:]
			exchange = ex
		}
	}
	// 2. 判断后缀: 600000.sh or AAPL.us
	if exchange == meta.UNKNOWN && len(pureCode) >= 3 && pureCode[len(pureCode)-3] == '.' {
		if ex, err := meta.ParseExchange(pureCode[len(pureCode)-2:]); err == nil && ex != meta.UNKNOWN {
			ticker = pureCode[:len(pureCode)-3]
			exchange = ex
		}
	}

	if exchange == meta.UNKNOWN {
		switch len(pureCode) {
		case 4:
			if isAllAlpha(pureCode) {
				return meta.Instrument{Exchange: meta.USA, Type: meta.InstrumentTypeStock, Ticker: pureCode}
			}
		case 5:
			if isAllDigit(pureCode) {
				return meta.Instrument{Exchange: meta.HKEX, Type: meta.InstrumentTypeStock, Ticker: pureCode}
			}
		case 6:
			// 1. 全局规则优先匹配
			cr := ticker_rules.MatchRule(pureCode, ticker_rules.GlobalRules())
			if cr.Exchange != meta.UNKNOWN {
				return meta.Instrument{Exchange: cr.Exchange, Type: cr.InstrumentType, Ticker: pureCode}
			}
			// 2.1 0, 159和3开头, 优先匹配深交所
			if strings.HasPrefix(pureCode, "0") || strings.HasPrefix(pureCode, "159") || strings.HasPrefix(pureCode, "3") {
				cr = ticker_rules.MatchRule(pureCode, ticker_rules.SzseRules())
				if cr.Exchange != meta.UNKNOWN {
					return meta.Instrument{Exchange: cr.Exchange, Type: cr.InstrumentType, Ticker: pureCode}
				}
			}
			// 2.2 6和5开头, 优先匹配上交所
			if strings.HasPrefix(pureCode, "6") || strings.HasPrefix(pureCode, "5") {
				cr = ticker_rules.MatchRule(pureCode, ticker_rules.SseRules())
				if cr.Exchange != meta.UNKNOWN {
					return meta.Instrument{Exchange: cr.Exchange, Type: cr.InstrumentType, Ticker: pureCode}
				}
			}
			// 2.3 匹配上交所
			cr = ticker_rules.MatchRule(pureCode, ticker_rules.SseRules())
			if cr.Exchange != meta.UNKNOWN {
				return meta.Instrument{Exchange: cr.Exchange, Type: cr.InstrumentType, Ticker: pureCode}
			}
			// 2.4 匹配深交所
			cr = ticker_rules.MatchRule(pureCode, ticker_rules.SzseRules())
			if cr.Exchange != meta.UNKNOWN {
				return meta.Instrument{Exchange: cr.Exchange, Type: cr.InstrumentType, Ticker: pureCode}
			}
			// 2.5 匹配北交所
			cr = ticker_rules.MatchRule(pureCode, ticker_rules.BseRules())
			if cr.Exchange != meta.UNKNOWN {
				return meta.Instrument{Exchange: cr.Exchange, Type: cr.InstrumentType, Ticker: pureCode}
			}
		default:
			exchange = meta.UNKNOWN
			typ = meta.InstrumentTypeUnknown
		}
	}

	// 3. 如果exchange是UNKNOWN, 则返回未知规则
	if exchange == meta.UNKNOWN {
		return meta.Instrument{Exchange: meta.UNKNOWN, Type: meta.InstrumentTypeUnknown}
	}

	if typ == meta.InstrumentTypeUnknown {
		var rules []ticker_rules.CodeRule
		switch exchange {
		case meta.SSE:
			rules = ticker_rules.SseRules()
		case meta.SZSE:
			rules = ticker_rules.SzseRules()
		case meta.BSE:
			rules = ticker_rules.BseRules()
		case meta.HKEX:
			rules = ticker_rules.HkexRules()
		case meta.USA:
			rules = ticker_rules.UsaRules()
		default:
			return meta.Instrument{Exchange: meta.UNKNOWN, Type: meta.InstrumentTypeUnknown}
		}
		cr := ticker_rules.MatchRule(ticker, rules)
		if cr.InstrumentType != meta.InstrumentTypeUnknown {
			return meta.Instrument{Exchange: cr.Exchange, Type: cr.InstrumentType, Ticker: ticker}
		}
		return meta.Instrument{Exchange: meta.UNKNOWN, Type: meta.InstrumentTypeUnknown}
	}
	return meta.Instrument{Exchange: exchange, Type: typ, Ticker: ticker}
}

// ============================================================
// CorrectSecurityCode — corrects the format of a security code.
// ============================================================

// CorrectSecurityCode corrects the format of a security code string.
// Supports prefix (sh600000), suffix (600000.sh, AAPL.us), and raw (600000) formats.
// Returns the corrected code or the original on failure.
func CorrectSecurityCode(code string) string {
	code = strings.TrimSpace(code)
	if code == "" {
		return code
	}

	// Already has correct form (lowercase prefix + digits, e.g., "sh600000")
	lower := strings.ToLower(code)
	if len(lower) >= 8 {
		prefix := lower[:2]
		if (prefix == "sh" || prefix == "sz" || prefix == "bj") && isAllDigits(lower[2:]) {
			return lower
		}
	}

	// Suffix form: e.g., "600000.sh" → "sh600000"
	if len(lower) > 3 && lower[len(lower)-3] == '.' {
		suffix := lower[len(lower)-2:]
		ticker := lower[:len(lower)-3]
		switch suffix {
		case "sh":
			return "sh" + ticker
		case "sz":
			return "sz" + ticker
		case "bj":
			return "bj" + ticker
		default:
			return lower // non-CN markets keep suffix form
		}
	}

	// Raw 6-digit code: infer exchange
	if len(lower) == 6 && isAllDigits(lower) {
		switch {
		case lower[0] == '6' || lower[0] == '5' || lower[0] == '9' || strings.HasPrefix(lower, "88"):
			return "sh" + lower
		default:
			return "sz" + lower
		}
	}

	// Raw 4-digit code (BSE)
	if len(lower) == 4 && isAllDigits(lower) {
		if lower[0] == '8' || lower[0] == '4' {
			return "bj" + lower
		}
	}

	// US-style: 4 uppercase letters → keep as-is or add .us
	if len(lower) <= 5 && isAllAlpha(lower) {
		return lower
	}

	return lower
}

func isAllDigits(s string) bool {
	for _, c := range s {
		if c < '0' || c > '9' {
			return false
		}
	}
	return true
}

func isAllAlpha(s string) bool {
	if s == "" {
		return false
	}
	for _, c := range s {
		if (c < 'a' || c > 'z') && (c < 'A' || c > 'Z') {
			return false
		}
	}
	return true
}

func isAllDigit(s string) bool {
	if s == "" {
		return false
	}
	for _, c := range s {
		if c < '0' || c > '9' {
			return false
		}
	}
	return true
}

File: quant1x/data/market.cpp
// Copyright (c) Quant1X <wangfengxy@sina.cn>.
// Licensed under the MIT License.
//
// market — 市场/证券代码识别与纠正, 与 Python data/market.py 对齐

#include "market.h"
#include "meta/ticker_rules/market_sse.h"
#include "meta/ticker_rules/market_szse.h"
#include "meta/ticker_rules/market_bse.h"
#include "meta/ticker_rules/market_hkex.h"
#include "meta/ticker_rules/market_usa.h"
#include <cctype>
#include <algorithm>

namespace data {

using meta::Exchange;
using meta::Instrument;
using meta::InstrumentType;
using meta::ticker_rules::match_rule;
using meta::ticker_rules::global_rules;

// ============================================================
// 常量: 前缀交易所标识集合
// ============================================================
static std::unordered_set<std::string> make_prefix_identifiers() {
    return {
        meta::exchange_identifier(Exchange::SSE),
        meta::exchange_identifier(Exchange::SZSE),
        meta::exchange_identifier(Exchange::BSE),
        meta::exchange_identifier(Exchange::HKEX),
    };
}

static std::unordered_set<std::string> make_all_identifiers() {
    return {
        meta::exchange_identifier(Exchange::SSE),
        meta::exchange_identifier(Exchange::SZSE),
        meta::exchange_identifier(Exchange::BSE),
        meta::exchange_identifier(Exchange::HKEX),
        meta::exchange_identifier(Exchange::HKFE),
        meta::exchange_identifier(Exchange::USA),
    };
}

static const std::unordered_set<std::string>& prefix_identifiers() {
    static std::unordered_set<std::string> s = make_prefix_identifiers();
    return s;
}

static const std::unordered_set<std::string>& all_identifiers() {
    static std::unordered_set<std::string> s = make_all_identifiers();
    return s;
}

// 辅助: 字符串转小写
static std::string to_lower(const std::string& s) {
    std::string r;
    for (char c : s) {
        r += static_cast<char>(::tolower(static_cast<unsigned char>(c)));
    }
    return r;
}

// 辅助: 判断字符串是否全为ASCII字母
static bool is_all_alpha(const std::string& s) {
    for (char c : s) {
        if (!::isalpha(static_cast<unsigned char>(c))) return false;
    }
    return !s.empty();
}

// 辅助: 判断字符串是否全为ASCII数字
static bool is_all_digit(const std::string& s) {
    for (char c : s) {
        if (!::isdigit(static_cast<unsigned char>(c))) return false;
    }
    return !s.empty();
}

// ============================================================
// detect_instrument_type_by_rule
// ============================================================
InstrumentType detect_instrument_type_by_rule(Exchange exchange, const std::string& code) {
    std::vector<meta::ticker_rules::CodeRule> rules;
    switch (exchange) {
        case Exchange::SSE:  rules = meta::ticker_rules::sse_rules(); break;
        case Exchange::SZSE: rules = meta::ticker_rules::szse_rules(); break;
        case Exchange::BSE:  rules = meta::ticker_rules::bse_rules(); break;
        case Exchange::HKEX: rules = meta::ticker_rules::hkex_rules(); break;
        case Exchange::USA:  rules = meta::ticker_rules::usa_rules(); break;
        default: return InstrumentType::Unknown;
    }
    auto cr = match_rule(code, rules);
    return cr.instrument_type;
}

// ============================================================
// detect_symbol
// ============================================================
Instrument detect_symbol(const std::string& input_str) {
    std::string s = input_str;
    // trim
    size_t start = s.find_first_not_of(" \t\n\r");
    size_t end   = s.find_last_not_of(" \t\n\r");
    if (start == std::string::npos) {
        return Instrument{}; // Unknown
    }
    s = s.substr(start, end - start + 1);
    s = to_lower(s);

    std::string pure_code = s;
    std::string ticker;
    Exchange exchange = Exchange::UNKNOWN;
    InstrumentType typ = InstrumentType::Unknown;

    // 1. 判断前缀: sh600000
    if (pure_code.size() >= 2) {
        std::string prefix = pure_code.substr(0, 2);
        if (prefix_identifiers().count(prefix)) {
            ticker = pure_code.substr(2);
            try {
                exchange = meta::exchange_from_abbr(prefix);
            } catch (...) {
                exchange = Exchange::UNKNOWN;
            }
        }
    }

    // 2. 判断后缀: 600000.sh or AAPL.us
    if (exchange == Exchange::UNKNOWN && pure_code.size() >= 3) {
        size_t dot_pos = pure_code.size() - 3;
        if (pure_code[dot_pos] == '.') {
            std::string suffix = pure_code.substr(dot_pos + 1);
            if (all_identifiers().count(suffix)) {
                ticker = pure_code.substr(0, dot_pos);
                try {
                    exchange = meta::exchange_from_abbr(suffix);
                } catch (...) {
                    exchange = Exchange::UNKNOWN;
                }
            }
        }
    }

    // 3. 纯数字或者字母(无显式前缀/后缀)
    if (exchange == Exchange::UNKNOWN) {
        size_t code_len = pure_code.size();
        switch (code_len) {
            case 4:
                if (is_all_alpha(pure_code)) {
                    exchange = Exchange::USA;
                    typ = InstrumentType::Stock;
                    return Instrument{exchange, typ, pure_code, "", 100, 2, 0, 0, "", ""};
                }
                break;
            case 5:
                if (is_all_digit(pure_code)) {
                    exchange = Exchange::HKEX;
                    typ = InstrumentType::Stock;
                    return Instrument{exchange, typ, pure_code, "", 100, 2, 0, 0, "", ""};
                }
                break;
            case 6: {
                // 3.1 全局规则优先匹配
                auto cr = match_rule(pure_code, global_rules());
                if (cr.exchange != Exchange::UNKNOWN) {
                    return Instrument{cr.exchange, cr.instrument_type, pure_code, "", 100, 2, 0, 0, "", ""};
                }

                // 3.2 按市场匹配规则
                // 3.2.1 0, 159和3开头, 优先匹配深交所
                if (pure_code[0] == '0' || pure_code.substr(0, 3) == "159" || pure_code[0] == '3') {
                    cr = match_rule(pure_code, meta::ticker_rules::szse_rules());
                    if (cr.exchange != Exchange::UNKNOWN) {
                        return Instrument{cr.exchange, cr.instrument_type, pure_code, "", 100, 2, 0, 0, "", ""};
                    }
                }
                // 3.2.2 6和5开头, 优先匹配上交所
                if (pure_code[0] == '6' || pure_code[0] == '5') {
                    cr = match_rule(pure_code, meta::ticker_rules::sse_rules());
                    if (cr.exchange != Exchange::UNKNOWN) {
                        return Instrument{cr.exchange, cr.instrument_type, pure_code, "", 100, 2, 0, 0, "", ""};
                    }
                }
                // 3.2.3 匹配上交所
                cr = match_rule(pure_code, meta::ticker_rules::sse_rules());
                if (cr.exchange != Exchange::UNKNOWN) {
                    return Instrument{cr.exchange, cr.instrument_type, pure_code, "", 100, 2, 0, 0, "", ""};
                }
                // 3.2.4 匹配深交所
                cr = match_rule(pure_code, meta::ticker_rules::szse_rules());
                if (cr.exchange != Exchange::UNKNOWN) {
                    return Instrument{cr.exchange, cr.instrument_type, pure_code, "", 100, 2, 0, 0, "", ""};
                }
                // 3.2.5 匹配北交所
                cr = match_rule(pure_code, meta::ticker_rules::bse_rules());
                if (cr.exchange != Exchange::UNKNOWN) {
                    return Instrument{cr.exchange, cr.instrument_type, pure_code, "", 100, 2, 0, 0, "", ""};
                }
                break;
            }
            default:
                return Instrument{}; // Unknown
        }
    }

    // 4. 如果exchange是UNKNOWN, 则返回未知
    if (exchange == Exchange::UNKNOWN) {
        return Instrument{};
    }

    // 5. 如果typ是Unknown, 按市场规则匹配
    if (typ == InstrumentType::Unknown) {
        std::vector<meta::ticker_rules::CodeRule> rules;
        switch (exchange) {
            case Exchange::SSE:  rules = meta::ticker_rules::sse_rules(); break;
            case Exchange::SZSE: rules = meta::ticker_rules::szse_rules(); break;
            case Exchange::BSE:  rules = meta::ticker_rules::bse_rules(); break;
            case Exchange::HKEX: rules = meta::ticker_rules::hkex_rules(); break;
            case Exchange::USA:  rules = meta::ticker_rules::usa_rules(); break;
            default: return Instrument{};
        }

        auto cr = match_rule(ticker, rules);
        if (cr.instrument_type != InstrumentType::Unknown) {
            return Instrument{cr.exchange, cr.instrument_type, ticker, "", 100, 2, 0, 0, "", ""};
        } else {
            return Instrument{};
        }
    } else {
        return Instrument{exchange, typ, ticker, "", 100, 2, 0, 0, "", ""};
    }
}

// ============================================================
// correct_security_code
// ============================================================
std::string correct_security_code(const std::string& code) {
    if (code.empty()) {
        return "";
    }

    auto inst = detect_symbol(code);
    if (inst.can_construct_symbol()) {
        return inst.symbol();
    }

    // 回退: 简单的 strip + detect 逻辑
    std::string s = to_lower(code);
    // trim
    size_t start = s.find_first_not_of(" \t\n\r");
    size_t end   = s.find_last_not_of(" \t\n\r");
    if (start == std::string::npos) return "";
    s = s.substr(start, end - start + 1);

    static const char* market_flags[] = {"sh", "sz", "bj", "hk", "us"};

    // 前缀形式
    for (const char* flag : market_flags) {
        std::string f(flag);
        if (s.size() > f.size() && s.substr(0, f.size()) == f) {
            return f + s.substr(f.size());
        }
    }

    // 后缀形式
    for (const char* flag : market_flags) {
        std::string suffix = std::string(".") + flag;
        if (s.size() > suffix.size() && s.substr(s.size() - suffix.size()) == suffix) {
            return std::string(flag) + s.substr(0, s.size() - suffix.size());
        }
    }

    // 纯数字代码推断
    if (s.size() == 6 && is_all_digit(s)) {
        if (s[0] == '6' || s[0] == '5' || s[0] == '9') {
            return "sh" + s;
        } else {
            return "sz" + s;
        }
    }

    return code;
}

} // namespace data
